Refactor AssignmentPlanning

justin0022 opened the pull request:
Resolves https://github.com/tl-its-umich-edu/my-learning-analytics/issues/641.

I tried to simplify this view as much as I could. The behaviour is mostly unchanged ~- the one change I did make was to remove the "save setting" checkbox, and make it so that when the assignment filter dropdown list is selected, the selected value is immediately saved to the backend, with a little snackbar message informing the user that their preference has been saved.~

I think this is a nicer overall experience for the user. In the original version, when the checkbox was clicked, the checkbox disappeared, which was a bit jarring. 

I'm opening this PR as a draft to get your thoughts on the change from checkbox to no checkbox.

justin0022 commented:
@jennlove-um I've added the checkbox back. The checkbox no longer disappears when users click on them unlike in the previous version - I can make the checkbox disappear if that's the desired behaviour. Unchecking the checkbox right now does not "undo" the save, so it may be confusing to users.

I've been looking at how other applications implement saving user settings, and noticed that many (but not all) applications simply save settings as users click around (for instance, here: [Github](https://github.com/settings/notifications) and [Google calendar](https://calendar.google.com/calendar/r/settings)). I personally think this is a nicer experience, but it doesn't impact the UI in a significant way.

I think going forward, as more views are developed and there are more settings for users to customize, we should consider moving the settings for all the views into one place, to offer a consistent and predictable way for users to set/modify their settings.

pushyamig commented:
@justin0022 
Few Observations:
1. Always fetch the assignment data after getting the defaults. based on what the user default based on that we get the assignment. Onload of assignment view I see that 2 calls are made which are redundant. 
![Screen Shot 2019-08-30 at 1 00 42 PM](https://user-images.githubusercontent.com/8579775/64038661-9f1a2f80-cb26-11e9-8b5f-9140ac854a40.png)
2. Regarding the Wording changes next to checkbox, I think we still want to keep it the way it was before. I have consulted with @jennlove-um  and said to keep it the same.

> Unchecking the checkbox right now does not "undo" the save, so it may be confusing to users.
I agree with you, @jennlove-um  what's your call?

pushyamig commented:
There is a bug in the checkbox action when user changes the default to say 2% the 'useState` is not remembering it but on the load of the view the API call returns the correct saved default value.

jennlove-um commented:
Hi Justin,  The default filter setting is one of those UI decisions that our research team who designed the view had very definite ideas about. So for now, we need to not change the way setting the default filter works.

When the student lands on the Assignment page, the default filter displays along with the text "My current setting" next to the drop down box. (no checkbox displays)

When the student selects a different filter value, the checkbox displays with the text "Remember my setting" to indicate that they can change the default value.

If the student clicks the checkbox to change the setting, the checkbox disappears and is replaced by the text "My current setting" to indicate that the filter default has been changed.

I'd be interested in thinking more about how we can move toward current design approaches that use things like the snackbar rather than checkboxes.  We would need to mock up and discuss that design change with the research team to get their input since they spent a good deal of time talking about and agreeing on the current design.

Thanks!

justin0022 commented:
Thanks @pushyamig & @jennlove-um. 

I'll make the saving flow/messaging the same as it was before - should I remove the snackbar for now as well? I suppose it is redundant to have the snackbar and a message beside the checkbox. 

GradeDistribution uses snackbar when users choose to hide/show their grade currently. For sake of consistency it may be worth keeping it for AssignmentPlanning.

jennlove-um commented:
Hi Justin - Sure, let's keep the snackbar to keep it all consistent.
Thanks!

On Fri, Aug 30, 2019 at 2:22 PM Justin Lee <notifications@github.com> wrote:

> Thanks @pushyamig <https://github.com/pushyamig> & @jennlove-um
> <https://github.com/jennlove-um>.
>
> I'll make the saving flow/messaging the same as it was before - should I
> remove the snackbar for now as well? I suppose it is redundant to have the
> snackbar and a message beside the checkbox.
>
> GradeDistribution uses snackbar when users choose to hide/show their grade
> currently. For sake of consistency it may be worth keeping it for
> AssignmentPlanning.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tl-its-umich-edu/my-learning-analytics/pull/676?email_source=notifications&email_token=AJOML4CLESHJTQC6S6NU7RDQHFQO3A5CNFSM4IQMMXBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5SNDGI#issuecomment-526700953>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJOML4CZJMJSLBQW63T2U33QHFQO3ANCNFSM4IQMMXBA>
> .
>


-- 
*Jennifer Love* (pronouns: she, her, hers <http://www.mypronouns.org>)
*Business Systems Analyst, ITS Teaching & Learning*
Information and Technology Services | University of Michigan
120-10 Boyer Building <https://its.umich.edu/about/visiting#boyer>, 201
East Hoover Avenue, Ann Arbor, MI 48019

*How can we help you?*
*For assistance with My Learning Analytics, *please contact
myla-help@umich.edu.
*For assistance with CTools,* please contact ctools-projectsites@umich.edu.

*For other topics, *contact the ITS Service Center
<https://its.umich.edu/help>:

   - Chat: chatsupport.it.umich.edu
   - Call: 734-764-HELP (764-4357)
   - Email: 4HELP@umich.edu

justin0022 commented:
I've reverted the saving flow back to the original style and made a few improvements:

1. The checkbox and "Remember my setting" text do not disappear/change until the setting is actually saved. In the original implementation, the checkbox disappeared and "My current setting" appeared _before_ the server responded with a 200 status.  
2. When the grade filter dropdown is changed, the table below shows a spinner while the data is being fetched. In the original implementation, the previous table was shown until the data was fetched. 

I also made sure that the assignment data is only fetched after the user setting loads, although my implementation is a bit hacky - I added a `fetchAfter` parameter to `useFetch`.

justin0022 commented:
> There is a bug in the checkbox action when user changes the default to say 2% the 'useState` is not remembering it but on the load of the view the API call returns the correct saved default value.

This issue should be fixed now @pushyamig

pushyamig commented on assets/src/containers/AssignmentPlanning.js:
```diff
@@ -65,130 +60,83 @@ const assignmentTable = assignmentData => {
+  const [saveSettingClicked, setSaveSettingClicked] = useState(false)
+
+  // this is the filter setting currently set
+  const [assignmentGradeFilter, setAssignmentGradeFilter] = useState(0)
```
```suggestion
  const [assignmentGradeFilter, setAssignmentGradeFilter] = useState('')
```
If user has not set any default at all the UI is just showing the spinner. the API call `http://localhost:5001/api/v1/courses/295532/get_user_default_selection?default_type=assignment` response in this case is {"default":""} L83-L85 correctly identifies this case but since the default in `useState(0)` and again we are setting with same 0 values with `setAssignmentGradeFilter(0)` so React Dom is not updating since it doesn't see any change is made. 

To replicated this try to delete the user default setting in the `user_default_selection` or pick a new user you haven't tesing add the user from the admin interface and become that user

pushyamig commented:
Other than that one suggested [change](https://github.com/tl-its-umich-edu/my-learning-analytics/pull/676/files#r321719828) the refactor looks Very Good!
